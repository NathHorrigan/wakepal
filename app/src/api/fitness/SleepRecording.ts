export enum SleepStage {
  Rem = 'REM',
  Deep = 'DEEP',
  Light = 'LIGHT',
  Awake = 'AWAKE',
}

export class MinuteSleepRecording {
  private stage: SleepStage
  private timestamp: Date

  constructor(timestamp: Date, stage: SleepStage) {
    this.stage = stage
    this.timestamp = timestamp
  }

  getStage(): SleepStage {
    return this.stage
  }

  getTimestamp(): Date {
    return this.timestamp
  }
}

export type NightsSleepRecording = Array<MinuteSleepRecording>

export abstract class BaseSleepRecording {
  private date: Date
  private recordings: NightsSleepRecording

  constructor(date: Date, recordings: NightsSleepRecording) {
    this.date = date
    this.recordings = recordings
  }

  getDate(): Date {
    return this.date
  }

  getRecordings(): NightsSleepRecording {
    return this.recordings
  }
}
