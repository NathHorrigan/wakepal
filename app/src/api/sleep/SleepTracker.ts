import { BaseSleepRecording } from './SleepRecording'

export interface SleepTracker {
  getLatestRecording(): BaseSleepRecording
  getRecordingAt(date: Date): BaseSleepRecording
  getRecordingsBetween(start: Date, end: Date): BaseSleepRecording[]
  getAllRecordings(start: Date, end: Date): BaseSleepRecording[]
}
