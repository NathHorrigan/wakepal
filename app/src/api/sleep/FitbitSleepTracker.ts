import { SleepTracker } from './SleepTracker'
import { BaseSleepRecording } from './SleepRecording'

export class FitbitSleepRecording extends BaseSleepRecording {}

export class FitbitSleepTracker implements SleepTracker {
  private static client: FitbitSleepTracker

  static getClient(): FitbitSleepTracker {
    if (!FitbitSleepTracker.client) {
      FitbitSleepTracker.client = new FitbitSleepTracker()
    }

    return FitbitSleepTracker.client
  }
}
