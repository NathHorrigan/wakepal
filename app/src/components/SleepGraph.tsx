import React, { SFC } from 'react'
import { Dimensions } from 'react-native'
import styled from 'styled-components/native'
import differenceInHours from 'date-fns/difference_in_hours'
import addHours from 'date-fns/add_hours'

import { SleepStage } from '@api/sleep/SleepRecording'
import { colors, fonts } from '@utils/theme'

interface SleepStageSegment {
  stage: SleepStage
  startTime: Date
  endTime: Date
}

interface SleepGraphProps {
  segments: SleepStageSegment[]
}

const SleepGraph: React.SFC<SleepGraphProps> = ({ segments }) => {
  const timespan = generateTimeline(segments)
  const stageColors = [
    colors.darkBlue,
    colors.blue,
    colors.paleBlue,
    colors.paleGreen,
  ]

  const sortSegment = (stage: SleepStage) =>
    segments.filter(segment => segment.stage === stage)

  const sortedSegements = {
    Rem: sortSegment(SleepStage.Rem),
    Deep: sortSegment(SleepStage.Deep),
    Light: sortSegment(SleepStage.Light),
    Awake: sortSegment(SleepStage.Awake),
  }

  const stages = Object.keys(sortedSegements)
  const { awakeHours, asleepHours, efficiency } = getSleepSummary(segments)

  return (
    <GraphContainer>
      <LabelRow>
        {stages.map((stage, index) => (
          <StageKey key={`${stage}-key`}>
            <StageDot color={stageColors[index]} />
            <StageLabel>{stage}</StageLabel>
          </StageKey>
        ))}
      </LabelRow>
      {stages.map((stage, index) => (
        <StageRow key={`${stage}-row`}>
          {sortedSegements[stage].map(segment => (
            <StageBar
              key={`${stage}-${index}-bar`}
              color={stageColors[index]}
              {...getSegementSpan(segment, timespan)}
            />
          ))}
          <StageTrace />
        </StageRow>
      ))}
      <Timeline>
        {timespan.map(hour => (
          <Hour key={`hour-${hour}`}>{hour}</Hour>
        ))}
      </Timeline>

      <SummaryContainer>
        <SleepSummary>
          <SummaryValueText color={colors.paleBlue}>
            {asleepHours}
          </SummaryValueText>
          <SummayLabelText>Hours Asleep</SummayLabelText>
        </SleepSummary>
        <SleepSummary>
          <SummaryValueText color={colors.paleGreen}>
            {awakeHours}
          </SummaryValueText>
          <SummayLabelText>Hours Awake</SummayLabelText>
        </SleepSummary>
        <SleepSummary>
          <SummaryValueText color={colors.coral}>{efficiency}</SummaryValueText>
          <SummayLabelText>Sleep Efficiency</SummayLabelText>
        </SleepSummary>
      </SummaryContainer>
    </GraphContainer>
  )
}

const GraphContainer = styled.View`
  max-width: 100%;
`

const LabelRow = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

const StageKey = styled(LabelRow)`
  margin: 20px 7px;
`

const StageDot = styled.View`
  height: 6px;
  width: 6px;
  border-radius: 7px;
  margin-right: 5px;
  background: ${props => props.color};
`

const StageLabel = styled.Text`
  color: #999999;
  font-size: 9px;
  font-family: ${fonts.medium};
`

const StageRow = styled.View`
  height: 7px;
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-bottom: 12px;
`

const StageTrace = styled.View`
  height: 1px;
  width: 100%;
  background: rgba(0, 0, 0, 0.08);
  z-index: -1;
`

const StageBar = styled.View`
  height: 7px;
  width: ${props => props.width}px;
  background: ${props => props.color};
  border-radius: 7px;
  position: absolute;

  left: ${props => props.left}px;
  right: ${props => props.right}px;
`

const Timeline = styled.View`
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-top: 2px;
  padding: 0 10px;
`

const Hour = styled.Text`
  font-size: 7px;
  font-family: ${fonts.bold};
  color: rgba(0, 0, 0, 0.4);
  text-align: center;
`

const SummaryContainer = styled.View`
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  margin-top: 35px;
`

const SleepSummary = styled.View`
  display: flex;
  align-items: center;
`

const SummaryValueText = styled.Text`
  font-size: 25px;
  font-family: ${fonts.bold};
  color: ${props => props.color};
  text-align: center;
`

const SummayLabelText = styled.Text`
  font-size: 11px;
  font-family: ${fonts.medium};
  color: #9e9e9e;
  text-align: center;
`

const getSleepSummary = (segments: SleepStageSegment[]) => {
  let segmentHours = segments.map(({ startTime, endTime, stage }) => ({
    stage,
    duration: differenceInHours(endTime, startTime),
  }))

  let calculateHours = filterFn =>
    segmentHours
      .filter(filterFn)
      .map(segment => segment.duration)
      .reduce((a, b) => a + b)

  const awakeHours = calculateHours(
    segment => segment.stage === SleepStage.Awake
  )
  const asleepHours = calculateHours(
    segment => segment.stage !== SleepStage.Awake
  )
  const efficiency = Math.ceil((asleepHours / (awakeHours + asleepHours)) * 100)

  return {
    awakeHours,
    asleepHours,
    efficiency,
  }
}

const getSegementSpan = (
  segment: SleepStageSegment,
  timeline: string[]
): number[] => {
  const { startTime, endTime } = segment
  const numTimeline = timeline.map(hour => Number(hour))

  // Calculate the positions of where each timeline starts
  const startHour = numTimeline.indexOf(startTime.getHours())
  const endHour = numTimeline.indexOf(endTime.getHours())
  const screenWidth = Dimensions.get('window').width
  const percentage = screenWidth / (numTimeline.length - 1)

  // Covert to percentage and add margins of 20px each side
  const startLeft = Math.max(10, startHour * percentage)
  const endRight = Math.min(screenWidth - 10, endHour * percentage)

  return {
    width: endRight - startLeft,
    left: startLeft,
    right: endRight,
  }
}

const generateTimeline = (segments: SleepStageSegment[]): string[] => {
  // Start and finish of sleep
  const startTime = segments[0].startTime
  const endTime = segments[segments.length - 1].endTime
  // How long their asleep
  const hourDiff = differenceInHours(endTime, startTime)
  // Generate array of hours, For example [19, 20, 21, ..., 09, 10]
  return Array(hourDiff + 1)
    .fill(0)
    .map((_, index) =>
      addHours(startTime, index)
        .getHours()
        .toLocaleString('en-US', {
          minimumIntegerDigits: 2,
          useGrouping: false,
        })
    )
}

export default SleepGraph
