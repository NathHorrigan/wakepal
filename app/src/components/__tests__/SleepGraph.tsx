import React from 'react'
import renderer from 'react-test-renderer'

import { SleepStage } from '../../api/sleep/SleepRecording'
import SleepGraph from '../SleepGraph'

it('renders correctly', () => {
  // Check button matches screenshot
  const tree = renderer
    .create(
      <SleepGraph
        segments={[
          {
            startTime: new Date(1585249742 * 1000),
            endTime: new Date(1585256952 * 1000),
            stage: SleepStage.Awake,
          },
          {
            startTime: new Date(1585256952 * 1000),
            endTime: new Date(1585264222 * 1000),
            stage: SleepStage.Light,
          },
          {
            startTime: new Date(1585264222 * 1000),
            endTime: new Date(1585278647 * 1000),
            stage: SleepStage.Rem,
          },
          {
            startTime: new Date(1585278647 * 1000),
            endTime: new Date(1585297232 * 1000),
            stage: SleepStage.Deep,
          },
        ]}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
