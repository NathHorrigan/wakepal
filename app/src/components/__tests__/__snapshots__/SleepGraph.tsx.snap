// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "maxWidth": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 20,
            "marginLeft": 7,
            "marginRight": 7,
            "marginTop": 20,
          },
        ]
      }
    >
      <View
        color="#035388"
        style={
          Array [
            Object {
              "backgroundColor": "#035388",
              "borderRadius": 7,
              "height": 6,
              "marginRight": 5,
              "width": 6,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 9,
            },
          ]
        }
      >
        Rem
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 20,
            "marginLeft": 7,
            "marginRight": 7,
            "marginTop": 20,
          },
        ]
      }
    >
      <View
        color="#2680C2"
        style={
          Array [
            Object {
              "backgroundColor": "#2680C2",
              "borderRadius": 7,
              "height": 6,
              "marginRight": 5,
              "width": 6,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 9,
            },
          ]
        }
      >
        Deep
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 20,
            "marginLeft": 7,
            "marginRight": 7,
            "marginTop": 20,
          },
        ]
      }
    >
      <View
        color="#62B0E8"
        style={
          Array [
            Object {
              "backgroundColor": "#62B0E8",
              "borderRadius": 7,
              "height": 6,
              "marginRight": 5,
              "width": 6,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 9,
            },
          ]
        }
      >
        Light
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 20,
            "marginLeft": 7,
            "marginRight": 7,
            "marginTop": 20,
          },
        ]
      }
    >
      <View
        color="#A3D9A5"
        style={
          Array [
            Object {
              "backgroundColor": "#A3D9A5",
              "borderRadius": 7,
              "height": 6,
              "marginRight": 5,
              "width": 6,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 9,
            },
          ]
        }
      >
        Awake
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": 7,
          "marginBottom": 12,
          "width": "100%",
        },
      ]
    }
  >
    <View
      color="#035388"
      left={230.76923076923077}
      right={461.53846153846155}
      style={
        Array [
          Object {
            "backgroundColor": "#035388",
            "borderRadius": 7,
            "height": 7,
            "left": 230.76923076923077,
            "position": "absolute",
            "right": 461.53846153846155,
            "width": 230.76923076923077,
          },
        ]
      }
      width={230.76923076923077}
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.08)",
            "height": 1,
            "width": "100%",
            "zIndex": -1,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": 7,
          "marginBottom": 12,
          "width": "100%",
        },
      ]
    }
  >
    <View
      color="#2680C2"
      left={461.53846153846155}
      right={740}
      style={
        Array [
          Object {
            "backgroundColor": "#2680C2",
            "borderRadius": 7,
            "height": 7,
            "left": 461.53846153846155,
            "position": "absolute",
            "right": 740,
            "width": 278.46153846153845,
          },
        ]
      }
      width={278.46153846153845}
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.08)",
            "height": 1,
            "width": "100%",
            "zIndex": -1,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": 7,
          "marginBottom": 12,
          "width": "100%",
        },
      ]
    }
  >
    <View
      color="#62B0E8"
      left={115.38461538461539}
      right={230.76923076923077}
      style={
        Array [
          Object {
            "backgroundColor": "#62B0E8",
            "borderRadius": 7,
            "height": 7,
            "left": 115.38461538461539,
            "position": "absolute",
            "right": 230.76923076923077,
            "width": 115.38461538461539,
          },
        ]
      }
      width={115.38461538461539}
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.08)",
            "height": 1,
            "width": "100%",
            "zIndex": -1,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": 7,
          "marginBottom": 12,
          "width": "100%",
        },
      ]
    }
  >
    <View
      color="#A3D9A5"
      left={10}
      right={115.38461538461539}
      style={
        Array [
          Object {
            "backgroundColor": "#A3D9A5",
            "borderRadius": 7,
            "height": 7,
            "left": 10,
            "position": "absolute",
            "right": 115.38461538461539,
            "width": 105.38461538461539,
          },
        ]
      }
      width={105.38461538461539}
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.08)",
            "height": 1,
            "width": "100%",
            "zIndex": -1,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 2,
          "paddingBottom": 0,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 0,
          "width": "100%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      19
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      20
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      21
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      22
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      23
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      00
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      01
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      02
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      03
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      04
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      05
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      06
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      07
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.4)",
            "fontFamily": "SFProRounded-Bold",
            "fontSize": 7,
            "textAlign": "center",
          },
        ]
      }
    >
      08
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginTop": 35,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
          },
        ]
      }
    >
      <Text
        color="#62B0E8"
        style={
          Array [
            Object {
              "color": "#62B0E8",
              "fontFamily": "SFProRounded-Bold",
              "fontSize": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        11
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#9e9e9e",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 11,
              "textAlign": "center",
            },
          ]
        }
      >
        Hours Asleep
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
          },
        ]
      }
    >
      <Text
        color="#A3D9A5"
        style={
          Array [
            Object {
              "color": "#A3D9A5",
              "fontFamily": "SFProRounded-Bold",
              "fontSize": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        2
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#9e9e9e",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 11,
              "textAlign": "center",
            },
          ]
        }
      >
        Hours Awake
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
          },
        ]
      }
    >
      <Text
        color="#F86A6A"
        style={
          Array [
            Object {
              "color": "#F86A6A",
              "fontFamily": "SFProRounded-Bold",
              "fontSize": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        85
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#9e9e9e",
              "fontFamily": "SFProRounded-Medium",
              "fontSize": 11,
              "textAlign": "center",
            },
          ]
        }
      >
        Sleep Efficiency
      </Text>
    </View>
  </View>
</View>
`;
